### PR TITLE
Allow passing filters when reading from an ixmp4 platform

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- ~Name of contributors Added to AUTHORS.rst~
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR allows passing filters when reading from an ixmp4 platform and it extends `read_iiasa()` to first try reading from ixmp4 platforms hosted by IIASA.

### CODE REVIEW

File: pyam/iiasa.py [hunk 1]
```diff
@@ -648,16 +660,24 @@ def lazy_read_iiasa(
...
+        Arguments for :meth:`pyam.read_ixmp4` or :meth:`pyam.iiasa.Connection.query`.
+
+    Notes
+    -----
+    This feature does currently not support reading data from |ixmp4| platforms.
+
+    Credentials (username & password) are not required to access any public |ixmp4|
+    or Scenario Explorer database (i.e., with Guest login).
     """
+    if name in [i.name for i in ixmp4.conf.settings.manager.list_platforms()]:
```
Comment (phackstock) [inline pyam/iiasa.py:677]: For better readability I always prefer to use the first letter of the object or better the full word:
```suggestion
    if name in [platform.name for platform in ixmp4.conf.settings.manager.list_platforms()]:
```

File: pyam/iiasa.py [hunk 2]
```diff
@@ -648,16 +660,24 @@ def lazy_read_iiasa(
...
+        raise NotImplementedError(
```
Comment (phackstock) [inline pyam/iiasa.py:678]: Could it be the case that a name of a Scenario Explorer and ixmp4 is the same and therefore would run into an error here?

File: pyam/iiasa.py [hunk 3]
```diff
@@ -589,31 +590,42 @@ def _new_default_api(kwargs):
...
+        Arguments for :meth:`pyam.read_ixmp4` or :meth:`pyam.iiasa.Connection.query`.
+
+    Notes
+    -----
+    Credentials (username & password) are not required to access any public |ixmp4|
+    or Scenario Explorer database (i.e., with Guest login).
     """
+    if name in [i.name for i in ixmp4.conf.settings.manager.list_platforms()]:
+        if meta is not True:
+            raise NotImplementedError(
```
Comment (phackstock) [inline pyam/iiasa.py:624]: If there are name overlaps between a Scenario Explorer and an ixmp4 platform I think we could run into an issue here. 
If a user wants `meta=True` but with a Scenario Explorer we could wrongly get an error here.

**Review by phackstock — COMMENTED** [review]
Looks good to me overall. 
Some minor comments in line.

File: pyam/iiasa.py [hunk 4]
```diff
@@ -589,31 +590,42 @@ def _new_default_api(kwargs):
 def read_iiasa(
     name, default_only=True, meta=True, creds=None, base_url=_AUTH_URL, **kwargs
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply pyam/iiasa.py:591]: Indeed, will remove as named argument.

File: pyam/ixmp4.py [hunk 5]
```diff
@@ -70,8 +104,7 @@ def write_to_ixmp4(platform: ixmp4.Platform | str, df):
         if missing := set(values).difference(platform_values):
             raise model.NotFound(
                 ", ".join(missing)
-                + f". Use `Platform.{dimension}.create()` to add the missing "
-                f"{dimension}."
+                + f". Use `Platform.{dimension}.create()` to add missing elements."
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply pyam/ixmp4.py:107]: The missing items are already printed via the line above, then a help-statement on how to add these element is written to the log.

**Comment by danielhuppmann** [comment]
Responding here to a larger comment by @phackstock in the review: this PR only touches the `pyam.read_iiasa()` method, and if there are ixmp databases and ixmp4 platforms with the same name, the default is now using ixmp4.

But a user will still have the possibility to make that selection explicitly either using:
```python
df_ixmp = pyam.iiasa.Connection(name).query(**filters)
df_ixmp4 = pyam.read_ixmp4(name, **filters)
```

It is our responsibility to keep potential overlaps of "database existence" to a minimum and keep the actual data consistent between new and legacy instances.

File: pyam/ixmp4.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (phackstock) [reply pyam/ixmp4.py:99]: Ah sorry never mind, I overlooked that.

**Comment by phackstock** [comment]
@danielhuppmann  sounds good to me.

File: pyam/ixmp4.py [hunk 6]
```diff
@@ -8,41 +8,75 @@
...
+    iamc_filters = _make_dict(
+        run=dict(default_only=default_only),
+        model=model,
+        scenario=scenario,
+        region=region,
+        variable=variable,
+        unit=unit,
+        year=year,
+    )
+    data = platform.iamc.tabulate(**iamc_filters)
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply pyam/ixmp4.py:52]: Removed

**Review by phackstock — APPROVED** [review]
Good to be merged from my side

**Pull request merged**